excel.go

```diff
@@ -0,0 +1,235 @@
+		if _, ok := macFinUsersToPasswords[pwUser]; !ok {
+			// pwUser is not in MACFin users; mark row for deletion from automatedSheet
+			rowsToDelete = append(rowsToDelete, pwRow.Row)
+			continue
+		}
+	}
+
+	// delete rows from automatedSheet that are marked for deletion
+	// iterate in descending order
+	sort.Ints(rowsToDelete)
+	for idx := len(rowsToDelete) - 1; idx >= 0; idx-- {
+		rowToDelete := toSheetCoord(rowsToDelete[idx] + rowOffset)
+		err := f.RemoveRow(automatedSheet, rowToDelete)
+		if err != nil {
+			return fmt.Errorf("failed removing row %d from %s sheet in file %s: %s", rowToDelete, automatedSheet, input.Filename, err)
+		}
+	}
+
+	err = f.Save()
+	if err != nil {
+		return fmt.Errorf("failed saving %s after synchronizing automated sheet users to MACFin users: %s", input.Filename, err)
+	}
+
+	return nil
+}
+
+func writeCell(f *excelize.File, sheet string, xCoord, yCoord int, value string) error {
+	cellName, err := excelize.CoordinatesToCellName(toSheetCoord(xCoord), toSheetCoord(yCoord))
+	if err != nil {
+		return err
+	}
+	err = f.SetCellValue(sheet, cellName, value)
+	if err != nil {
+		return err
+	}
+	err = f.Save()
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Copy cell in automatedSheet
+func copyCell(f *excelize.File, automatedSheetName string, srcX, srcY, destX, destY int) error {
+	srcCell, err := excelize.CoordinatesToCellName(toSheetCoord(srcX), toSheetCoord(srcY))
+	if err != nil {
+		return err
+	}
+	destCell, err := excelize.CoordinatesToCellName(toSheetCoord(destX), toSheetCoord(destY))
+	if err != nil {
+		return err
+	}
+	srcValue, err := f.GetCellValue(automatedSheetName, srcCell)
+	if err != nil {
+		return err
+	}
+	err = f.SetCellValue(automatedSheetName, destCell, srcValue)
+	if err != nil {
+		return err
+	}
+	err = f.Save()
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func validateFileSize(f *excelize.File, input *Input) (errors []string) {
+	v := new(Validator)
+
+	rows, err := f.GetRows(input.SheetName)
+	if err != nil {
+		v.Errorf("Error getting rows: %s", err)
+
+	}
+	if len(rows) > excelize.TotalRows {
+		v.Errorf("Number of rows %d exceeds max number %d", len(rows), excelize.TotalRows)
+	}
+
+	if len(rows[0]) > excelize.TotalColumns {
+		v.Errorf("Number of columns %d exceeds max number %d", len(rows[0]), excelize.TotalColumns)
+	}
+
+	return v.Errors
+}
+
+// Write new password to password column in the testing sheet
+func updateMACFinUsers(f *excelize.File, input *Input) error {
+	userToPasswordRow, err := getManagedUsers(f, input)
+	if err != nil {
+		return err
+	}
+
+	rows, err := f.GetRows(input.SheetName)
+	if err != nil {
+		return err
+	}
+
+	numPasswordsUpdated := 0
+	headerNameToXCoord := getHeaderToXCoord(rows[0])
+	userX := headerNameToXCoord[input.UsernameHeader]
+	passwordX := headerNameToXCoord[input.PasswordHeader]
+	rowOffset := input.RowOffset
+
+	for i, row := range rows[rowOffset:] {
+		if row == nil {
```

Comment:
when does this happen?

Reply:
I make this check when I read the MACFin sheet (lines 42 and 216 in excel.go). If the MACFin sheet has a blank row, and I remove this check, my code will panic on lines 45 and 219.

Reply:
Oh I see.

Seems like you should check that the length of the row is greater than `userX` and `passwordX`, no?

Reply:
Yes. len(row) cannot be less than the max of userX and passwordX. lines 42 and 216. Done.